### ISSUE TITLE
circular dependency

### ISSUE THREAD

**Issue by jhgoebbert** [mbdevpl/python-boilerplates#13]
Hi,
I came across this circular dependency:
`boilerplates` needs `version-query` -> https://github.com/mbdevpl/python-boilerplates/blob/v1.1.1/requirements.txt
and
 `version-query` needs `boilerplates`  -> https://github.com/mbdevpl/version-query/blob/v1.5.5/requirements.txt
while building [tvb-ext-bucket](https://github.com/the-virtual-brain/tvb-ext-bucket) from source.

**Comment by adityaraute** [comment]
I wonder if there's a way to resolve this at all

**Comment by mbdevpl** [comment]
@jhgoebbert Hello! Is this dependency causing any particular issue that you can point to? Such as:

I run this command (involving boilerplates and/or version-query):
```
(...)
```

and expect this (good result):
```
(...)
```

but I get this (error that mentions the circular dependency):
```
(...)
```

I'm asking because I use `version-query` as well as `boilerplates` a lot and I can install and use both without any issues. Also I maintain several Python packages that depend on both extensively and they seem to install and run just fine. Examples:

* https://github.com/mbdevpl/ingit
* https://github.com/mbdevpl/system-query
* https://github.com/mbdevpl/timing
* https://github.com/mbdevpl/argunparse

**Comment by jhgoebbert** [comment]
We are building our software from source. Perhaps you use a prebuild wheel.
But the requirements-files are kind of clear, aren't they? It is a circular dependency which is never a good idea.

**Comment by mbdevpl** [comment]
Yes, requirements files are clear, however, when building Python packages (from source or otherwise), requirements files do not matter at all as they specify runtime requirements, not build requirements. Build requirements are specified in [build-system] section of pyprojects.toml.

Also, when building _your_ software from source, I presume you are able to use prebuilt wheels of _other_ software such as python-boilerplates and/or verion-query?

I do get your comment about such dependency not being a good idea generally, however, is there any problem with giving specific answers to my earlier question?

**Comment by jhgoebbert** [comment]
Thank you for your comment. You are absolutely right, circular dependencies are only critical for build requirements.  
I will try to build it again and keep my fingers crossed it works.

**Comment by mbdevpl** [comment]
If you encounter an error again, I'd appreciate if you could post here how to reproduce it.

I was today experimenting with this, and inside a clean container (`docker run -it python:3.13 bash`) I was able to cause some error by running this:

```
pip install --no-cache-dir --no-binary version-query version-query==1.6.0
```

which spews a long log that ends in
```
        File "/usr/local/lib/python3.13/site-packages/pip/_internal/operations/build/build_tracker.py", line 102, in add
          raise LookupError(message)
      LookupError: https://files.pythonhosted.org/packages/25/2b/f92492fe576c7426afda1c52d1ee63fbdad53d29511bc3358824e35f729e/boilerplates-1.1.1.tar.gz (from https://pypi.org/simple/boilerplates/) (requires-python:>=3.8) is already being built: boilerplates from https://files.pythonhosted.org/packages/25/2b/f92492fe576c7426afda1c52d1ee63fbdad53d29511bc3358824e35f729e/boilerplates-1.1.1.tar.gz
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.

[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 2
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

So yes, indeed there's some issue, but without logs etc. I can't be sure if that's the same error you saw.
